### PR TITLE
feat: update the whole site for alpha 

### PR DESCRIPTION
**Pull request by Wilbert-mad**
remodel the whole site.
- [X] Tailwind added
- [x] Headless UI

<details open>
<summary>Routing</summary>
<br>

# Front-end Routes.

* [ ]  / => Home (display top rated users *as of now)
* [ ] /staff => Dmod staff list
* [x]  /search => search for some server or tags server (“?tags=1,2,3” or “?q=some%20server”)
Also can work with both (“?tags=bot&q=zero”)
* [x]  /servers => Display all the server this user can administrator in (Manage_Guild perms)
* [x] -   /servers/[guildID] => The main panel/view for everyone if fully completed otherwise only can be accessed by the server admins
* [x] -   /settings => Server settings
* [ ] -   /apps => View people that have applied
* [ ] -   /apps/[userID] => View application of user and process
* [ ] -   /apply => Make an application to this server, most likely use quarry to know the step (“?s=1”)
* [X] /[userVanity(default userId)] => The profile of a different user or this user (nice)
* [ ] /account=>Some type of hub for other user configs/settings?
* [x] -  /settings =>  Logged in user profile settings
* [ ] /admin
* [ ] -  /servers => list of all servers
* [ ] - -  /[guildID] => settings and overview on server
* [ ] - - /apps => this servers applications
* [ ] - - - /[userID] => application of this user to this server
* [ ] /user => list of users
* [ ] - - /[userID] => data of this user (include server this user has applied to ?)

# API Routes.

* /api
* [x] - /auth
* [X] - - /login => Discord login redirect
* [X] - - /redirect => Processing redirect
* [X] - - /current_user => User data logged in right now
* [X] - - /logout => Log out, destroy session
*  [x] - - /update => Update the user profile
* [x] /discord
* [x] - - /guilds => An array of [preview like guilds](https://discord.com/developers/docs/resources/guild#guild-preview-object) that the user can manage/owner
* [x] /server/[guildID]
* [x] - - /settings => (“POST”,“PATCH” methods) manage settings of server
* [ ] - - /apps => Return data on applications
* [ ] - - -  /[ID] => get,reject,update app or something to the application
</details>

<details open>
<summary>TODO</summary>

* [X] Vanity working
* [x] User able to update profile
* [x] Don't prompt the user to update their data DO IT OF THEM!

</details>

# Database Shema
Shema
<small>Will likely change a bit...</small>

### CODE REVIEW

**Comment by MattA-Official** [comment]
I will review this in a few days. Sorry for the delay; I'm on holiday.

**Review by MattA-Official — APPROVED** [review]

File: .example.env [hunk 1]
```diff
@@ -1,8 +1,13 @@
 SECRET_SESSION_PASSCODE=
 CLIENT_ID=
 CLIENT_SECRET=
+CLIENT_TOKEN=
 REDIRECT_URL=http://localhost:3000/api/auth/redirect
+INVITE_REDIRECT_URL=http://localhost:3000/api/auth/invite_redirect
 ENCRIPT_KEY=
```
Comment (alanlichen) [inline .example.env:7]: Was this meant to be encript or encrypt?

Comment (Wilbert-mad) [reply .example.env:7]: That is a typo yes, Thanks for catching it!

As well as `GUILD_ENCRIPT_KEY` but that will be removed.

Comment (alanlichen) [reply .example.env:7]: I was just skimming through the code, good to know that I can contribute to something that I have nothing to do with. Looking forward to the release :)

**Comment by rpxs** [comment]
Instead of putting tasks in a pull request, please add them to a new Project, which will make it easier for everyone to see what needs to be done.

File: components/guildStatus.tsx [hunk 2]
```diff
@@ -0,0 +1,123 @@
+import { clsx, PERMIUM_TIER, VERIFACTION_LEVEL } from '@lib/constants';
```
Comment (MattA-Official) [inline components/guildStatus.tsx:1]:
```suggestion
import { clsx, PREMIUM_TIER, VERIFICATION_LEVEL } from 'lib/constants';
```

File: components/guildStatus.tsx [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+          <b>Tags</b>: <span className='space-x-1'>{toLinkTags(['roblox', 'gaming'])}</span>
+          {/* <b>Tags</b>: <span className='space-x-1'>{guild.tags.length > 0 ? toLinkTags(guild.tags) : 'None'}</span> */}
+        </span>
+        <br />
+        <span title='Server looking for'>
+          <b>Looking for</b>: {ProperWithAnd(guild.look_types)}
+        </span>
+        <br />
+        <span title='Server premium tier'>
+          <b>Premium Tier</b>: {PERMIUM_TIER[guild.premium_tier]}
```
Comment (MattA-Official) [inline components/guildStatus.tsx:78]:
```suggestion
          <b>Premium Tier</b>: {PREMIUM_TIER[guild.premium_tier]}
```

File: components/guildStatus.tsx [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        </span>
+        <br />
+        <span title='Server verifction level'>
+          <b>Verfiction Level</b>: {VERIFACTION_LEVEL[guild.verification_level]}
```
Comment (MattA-Official) [inline components/guildStatus.tsx:82]:
```suggestion
          <b>Verfiction Level</b>: {VERIFICATION_LEVEL[guild.verification_level]}
```

File: components/profile.tsx [hunk 5]
```diff
@@ -0,0 +1,78 @@
+import MarkDown from '@lib/markdown';
+import { bannerReslover, clsx, evalBadges, evalFlags } from 'lib/constants';
```
Comment (MattA-Official) [inline components/profile.tsx:2]:
```suggestion
import { bannerResolver, clsx, evalBadges, evalFlags } from 'lib/constants';
```

File: components/profile.tsx [hunk 6]
```diff
@@ -0,0 +1,78 @@
+import MarkDown from '@lib/markdown';
```
Comment (MattA-Official) [inline components/profile.tsx:1]:
```suggestion
import MarkDown from 'lib/markdown';
```

File: components/guildView.tsx [hunk 7]
```diff
@@ -0,0 +1,149 @@
+import { clsx } from '@lib/constants';
```
Comment (MattA-Official) [inline components/guildView.tsx:1]:
```suggestion
import { clsx } from 'lib/constants';
```

File: components/guildView.tsx [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+import MarkDown from '@lib/markdown';
```
Comment (MattA-Official) [inline components/guildView.tsx:2]:
```suggestion
import MarkDown from 'lib/markdown';
```

File: components/guildView.tsx [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+import { GuildData } from '@models/guilds';
```
Comment (MattA-Official) [inline components/guildView.tsx:3]:
```suggestion
import { GuildData } from 'models/guilds';
```

File: components/guildStatus.tsx [hunk 10]
```diff
@@ -0,0 +1,123 @@
+import { clsx, PERMIUM_TIER, VERIFACTION_LEVEL } from '@lib/constants';
+import { GuildData } from '@models/guilds';
```
Comment (MattA-Official) [inline components/guildStatus.tsx:2]:
```suggestion
import { GuildData } from 'models/guilds';
```

File: lib/constants.tsx [hunk 11]
```diff
@@ -0,0 +1,153 @@
+export const user_flags = {
+  STAFF: 1 << 0, // 1
+  EARLY_SUPPORTER: 1 << 1, // 2
+  VERIFYED: 1 << 2, // 4
```
Comment (MattA-Official) [inline lib/constants.tsx:4]:
```suggestion
  VERIFIED: 1 << 2, // 4
```

File: lib/constants.tsx [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    </svg>
+  ),
+
+  // SITE_MOD: <></>,
+  // DEVELOPER: <></>,
+  // EARLY_SUPPORTER: <></>,
+  // MOD: <></>,
+};
+
+export const PERMIUM_TIER = ['None', 'Tier 1', 'Tier 2', 'Tier 3'];
```
Comment (MattA-Official) [inline lib/constants.tsx:64]:
```suggestion
export const PREMIUM_TIER = ['None', 'Tier 1', 'Tier 2', 'Tier 3'];
```

File: lib/constants.tsx [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+export const VERIFACTION_LEVEL = ['None', 'Low', 'Medium', 'High', 'Vary High'];
```
Comment (MattA-Official) [inline lib/constants.tsx:65]:
```suggestion
export const VERIFICATION_LEVEL = ['None', 'Low', 'Medium', 'High', 'Very High'];
```

File: lib/serverUpdateValidators.ts [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+  description({ value }) {
+    if (value.length >= DESCRIPTION_MIN && value.length <= DESCRIPTION_MAX_DATA.NORMAL) return { error: false };
+
+    return {
+      error: true,
+      message: 'Description invalid length',
+    };
+  },
+  short_description({ value }) {
+    if (value.length >= SHOR_DESCRIPTION_MIN && value.length <= SHOR_DESCRIPTION_MAX_DATA.NORMAL) return { error: false };
```
Comment (MattA-Official) [inline lib/serverUpdateValidators.ts:30]:
```suggestion
    if (value.length >= SHORT_DESCRIPTION_MIN && value.length <= SHORT_DESCRIPTION_MAX_DATA.NORMAL) return { error: false };
```

File: lib/serverUpdateValidators.ts [hunk 15]
```diff
@@ -0,0 +1,72 @@
+export const DESCRIPTION_MAX_DATA = { PREMIUM: 4000, NORMAL: 2000 };
+export const SHOR_DESCRIPTION_MAX_DATA = { PREMIUM: 200, NORMAL: 100 };
```
Comment (MattA-Official) [inline lib/serverUpdateValidators.ts:2]:
```suggestion
export const SHORT_DESCRIPTION_MAX_DATA = { PREMIUM: 200, NORMAL: 100 };
```

File: lib/serverUpdateValidators.ts [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+export const SHOR_DESCRIPTION_MIN = 20;
```
Comment (MattA-Official) [inline lib/serverUpdateValidators.ts:4]:
```suggestion
export const SHORT_DESCRIPTION_MIN = 20;
```

File: lib/userUpdateValidators.ts [hunk 17]
```diff
@@ -0,0 +1,139 @@
+import type { userData } from '@models/users';
```
Comment (MattA-Official) [inline lib/userUpdateValidators.ts:1]:
```suggestion
import type { userData } from 'models/users';
```

File: pages/account/settings.tsx [hunk 18]
```diff
@@ -0,0 +1,274 @@
+import Profile from '@components/profile';
```
Comment (MattA-Official) [inline pages/account/settings.tsx:1]:
```suggestion
import Profile from 'components/profile';
```

File: pages/account/settings.tsx [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+import { Switch } from '@headlessui/react';
+import { validators } from '@lib/userUpdateValidators';
```
Comment (MattA-Official) [inline pages/account/settings.tsx:3]:
```suggestion
import { validators } from 'lib/userUpdateValidators';
```

File: pages/api/auth/updates.ts [hunk 20]
```diff
@@ -0,0 +1,121 @@
+import { typeValidators, validators } from '@lib/userUpdateValidators';
```
Comment (MattA-Official) [inline pages/api/auth/updates.ts:1]:
```suggestion
import { typeValidators, validators } from 'lib/userUpdateValidators';
```

File: pages/index.tsx [hunk 21]
```diff
@@ -1,28 +1,65 @@
-import home from "../styles/home.module.scss";
-import MetaTags from "../components/MetaTags";
+import { clsx } from '@lib/constants';
```
Comment (MattA-Official) [inline pages/index.tsx:1]:
```suggestion
import { clsx } from 'lib/constants';
```

File: pages/index.tsx [hunk 22]
```diff
@@ -1,28 +1,65 @@
...
+import { PreviewGuildData } from '@models/preview_guilds';
```
Comment (MattA-Official) [inline pages/index.tsx:2]:
```suggestion
import { PreviewGuildData } from 'models/preview_guilds';
```

File: pages/search.tsx [hunk 23]
```diff
@@ -0,0 +1,148 @@
+import Layout from '@components/layout';
+import { PreviewGuildData } from '@models/preview_guilds';
+import { userData } from '@models/users';
```
Comment (MattA-Official) [inline pages/search.tsx:3]:
```suggestion
import Layout from 'components/layout';
import { PreviewGuildData } from 'models/preview_guilds';
import { userData } from 'models/users';
```

File: pages/api/v1/kei/search_u.ts [hunk 24]
```diff
@@ -0,0 +1,40 @@
+import userModule from '@models/users';
```
Comment (MattA-Official) [inline pages/api/v1/kei/search_u.ts:1]:
```suggestion
import userModule from 'models/users';
```

File: pages/api/v1/kei/search_g.ts [hunk 25]
```diff
@@ -0,0 +1,44 @@
+import GuildModule from '@models/guilds';
+import PreviewGuildModule from '@models/preview_guilds';
```
Comment (MattA-Official) [inline pages/api/v1/kei/search_g.ts:2]:
```suggestion
import GuildModule from 'models/guilds';
import PreviewGuildModule from 'models/preview_guilds';
```

File: pages/api/v1/servers/[guildID]/settings.ts [hunk 26]
```diff
@@ -0,0 +1,96 @@
+import { resolveGuildMemberPerms } from '@lib/backend-utils';
+import { user_flags } from '@lib/constants';
+import connectToDatabase from '@lib/mongodb.connection';
+import { typeValidators, validators } from '@lib/serverUpdateValidators';
+import GuildModule from '@models/guilds';
+import PreviewGuildModule from '@models/preview_guilds';
+import userModule from '@models/users';
```
Comment (MattA-Official) [inline pages/api/v1/servers/[guildID]/settings.ts:7]:
```suggestion
import { resolveGuildMemberPerms } from 'lib/backend-utils';
import { user_flags } from 'lib/constants';
import connectToDatabase from 'lib/mongodb.connection';
import { typeValidators, validators } from 'lib/serverUpdateValidators';
import GuildModule from 'models/guilds';
import PreviewGuildModule from 'models/preview_guilds';
import userModule from 'models/users';
```

File: pages/servers/[guildID]/index.tsx [hunk 27]
```diff
@@ -0,0 +1,75 @@
+import GuildView from '@components/guildView';
```
Comment (MattA-Official) [inline pages/servers/[guildID]/index.tsx:1]:
```suggestion
import GuildView from 'components/guildView';
```

File: pages/servers/[guildID]/settings.tsx [hunk 28]
```diff
@@ -0,0 +1,309 @@
...
+import Layout from '@components/layout';
+import { Switch } from '@headlessui/react';
+import { resolveGuildMemberPerms } from '@lib/backend-utils';
+import { clsx } from '@lib/constants';
+import connectToDatabase from '@lib/mongodb.connection';
+import { SHOR_DESCRIPTION_MAX_DATA, SHOR_DESCRIPTION_MIN } from '@lib/serverUpdateValidators';
+import withSession from '@lib/session';
+import { DESCRIPTION_MAX_DATA, DESCRIPTION_MIN } from '@lib/userUpdateValidators';
+import guilds, { GuildData } from '@models/guilds';
+import userModule, { userData } from '@models/users';
```
Comment (MattA-Official) [inline pages/servers/[guildID]/settings.tsx:11]:
```suggestion
import GuildView from 'components/guildView';
import Layout from 'components/layout';
import { Switch } from '@headlessui/react';
import { resolveGuildMemberPerms } from 'lib/backend-utils';
import { clsx } from 'lib/constants';
import connectToDatabase from 'lib/mongodb.connection';
import { SHORT_DESCRIPTION_MAX_DATA, SHORT_DESCRIPTION_MIN } from 'lib/serverUpdateValidators';
import withSession from 'lib/session';
import { DESCRIPTION_MAX_DATA, DESCRIPTION_MIN } from 'lib/userUpdateValidators';
import guilds, { GuildData } from 'models/guilds';
import userModule, { userData } from 'models/users';
```

**Review by MattA-Official — CHANGES_REQUESTED** [review]

File: components/guildStatus.tsx (hunk 4 above)
Comment (GamerCreator1) [reply components/guildStatus.tsx:82]: Verfiction Level should be changed to Verification Level

File: components/profile.tsx [hunk 29]
```diff
@@ -1,4 +1,4 @@
-import { evalBadges, evalFlags } from 'lib/constants';
+import { bannerReslover, evalBadges, evalFlags } from 'lib/constants';
```
Comment (GamerCreator1) [inline components/profile.tsx]:
```suggestion
import { bannerResolver, evalBadges, evalFlags } from 'lib/constants';
```

File: components/profile.tsx [hunk 30]
```diff
@@ -10,21 +10,35 @@ export default function Profile({ profile }: props) {
   const evaledFlags = evalFlags(profile.site_flags);
   const badges = evalBadges(evaledFlags);
 
+  const bannerData = bannerReslover(profile.banner);
```
Comment (GamerCreator1) [inline components/profile.tsx]:
```suggestion
  const bannerData = bannerResolver(profile.banner);
```

File: lib/constants.tsx [hunk 31]
```diff
@@ -74,3 +76,39 @@ export function evalBadges(resoloved: { [key: string]: boolean }): [JSX.Element,
...
+
+export type bannerTypes = 'color' | 'img' | 'unknown';
+
+export interface reslovedBanner {
+  type: bannerTypes;
+  color: string | null;
+  image: string | null;
+}
+
+export const bannerResloverRegExps = {
```
Comment (GamerCreator1) [inline lib/constants.tsx:91]:
```suggestion
export const bannerResolverRegExps = {
```

File: lib/constants.tsx [hunk 32]
```diff
@@ -74,3 +76,39 @@ export function evalBadges(resoloved: { [key: string]: boolean }): [JSX.Element,
...
+  COLOR_RGB: /^rgb\((\d{1,3}),(\d{1,3}),(\d{1,3})\)$/,
+  COLOR_HEX: /^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/,
+  IMAGE: /^https:\/\//,
+};
+
+/**
+ * * Color banner: "color:rgb()" or "color:#hex"
+ * * Image banner: "img:https:/--"
+ */
+export function bannerReslover(bannerResolvable: string): reslovedBanner {
```
Comment (GamerCreator1) [inline lib/constants.tsx]:
```suggestion
export function bannerResolver(bannerResolvable: string): resolvedBanner {
```

File: lib/constants.tsx [hunk 33]
```diff
@@ -74,3 +76,39 @@ export function evalBadges(resoloved: { [key: string]: boolean }): [JSX.Element,
...
+  const type = bannerResolvable.slice(0, bannerResolvable.indexOf(':'));
+  const data = bannerResolvable.slice(bannerResolvable.indexOf(':') + 1);
+  if (!['color', 'img'].includes(type)) return { type: 'unknown', color: null, image: null };
+
+  if (type === 'img') {
+    if (bannerResloverRegExps.IMAGE.test(data)) return { type: 'img', color: null, image: data };
```
Comment (GamerCreator1) [inline lib/constants.tsx:108]:
```suggestion
    if (bannerResolverRegExps.IMAGE.test(data)) return { type: 'img', color: null, image: data };
```

File: lib/constants.tsx [hunk 34]
```diff
@@ -74,3 +76,39 @@ export function evalBadges(resoloved: { [key: string]: boolean }): [JSX.Element,
...
+    return { type: 'unknown', color: null, image: null };
+  }
+
+  if (type === 'color') {
+    if (!bannerResloverRegExps.COLOR_HEX.test(data) && !bannerResloverRegExps.COLOR_RGB.test(data)) return { type: 'color', color: DEFAULT_BANNER_COLOR, image: null };
```
Comment (GamerCreator1) [inline lib/constants.tsx:113]:
```suggestion
if (!bannerResolverRegExps.COLOR_HEX.test(data) && !bannerResolverRegExps.COLOR_RGB.test(data))
    ```

File: pages/servers/index.tsx [hunk 35]
```diff
@@ -0,0 +1,111 @@
...
+                  className='flex rounded-md bg-servercard py-2 px-4 cursor-pointer space-x-2 max-w-2xl'
+                  onClick={() => {
+                    setGuildModleData(guild);
+                    setModalOpen(true);
+                  }}
+                >
+                  <img
+                    draggable={false}
+                    className='rounded-full w-14 h-14 md:w-20 md:h-20'
+                    src={`https://cdn.discordapp.com/icons/${guild.id}/${guild.icon}.png?size=128`}
```
Comment (GamerCreator1) [inline pages/servers/index.tsx:98]: If server does not have server icon, use dmod logo
```suggestion
                    src={
                      guild.icon == null
                        ? 'https://dmod.gg/logo.png'
                        : `https://cdn.discordapp.com/icons/${guild.id}/${guild.icon}.png?size=128`
                    }
```

File: pages/servers/index.tsx [hunk 36]
```diff
@@ -0,0 +1,111 @@
...
+          <div className='space-y-3 text-xl'>
+            <h1 className='text-center'>Guilds Applications found.</h1>
+            <div className='flex justify-center overflow-x-hidden'>
+              <div className='flex flex-col sm:grid grid-cols-3 gap-3 w-full sm:w-2/3'>
+                {userGuildData.included.map(guild => (
+                  <div key={guild.id} className='flex rounded-md bg-servercard py-2 px-4 cursor-pointer space-x-2 max-w-2xl' onClick={() => router.push(`/servers/${guild.id}`)}>
+                    <img
+                      draggable={false}
+                      className='rounded-full w-14 h-14 md:w-20 md:h-20'
+                      src={`https://cdn.discordapp.com/icons/${guild.id}/${guild.icon}.png?size=128`}
```
Comment (GamerCreator1) [inline pages/servers/index.tsx:71]: If server does not have server icon, use dmod logo
```suggestion
                    src={
                      guild.icon == null
                        ? 'https://dmod.gg/logo.png'
                        : `https://cdn.discordapp.com/icons/${guild.id}/${guild.icon}.png?size=128`
                    }
```

Comment (GamerCreator1) [inline pages/servers/index.tsx:71]:
```suggestion
                    src={
                      guild.icon == null
                        ? 'https://dmod.gg/logo.png'
                        : `https://cdn.discordapp.com/icons/${guild.id}/${guild.icon}.png?size=128`
                    }
```

File: pages/servers/index.tsx (hunk 35 above)
Comment (GamerCreator1) [inline pages/servers/index.tsx:98]:
```suggestion
                    src={
                      guild.icon == null
                        ? 'https://dmod.gg/logo.png'
                        : `https://cdn.discordapp.com/icons/${guild.id}/${guild.icon}.png?size=128`
                    }
```

**Review by GamerCreator1 — COMMENTED** [review]
If guild icon does not exist, default to dmod logo

**Pull request merged**